Add ability to read math in PDF documents

NSoiffer opened the pull request:
<!-- Please read and fill in the following template, for an explanation of the sections see:
https://github.com/nvaccess/nvda/blob/master/projectDocs/dev/githubPullRequestTemplateExplanationAndExamples.md
Please also note that the NVDA project has a Citizen and Contributor Code of Conduct which can be found at https://github.com/nvaccess/nvda/blob/master/CODE_OF_CONDUCT.md. NV Access expects that all contributors and other community members read and abide by the rules set out in this document while participating or contributing to this project. This includes creating or commenting on issues and pull requests.

Please initially open PRs as a draft.
When you would like a review, mark the PR as "ready for review".
See https://github.com/nvaccess/nvda/blob/master/.github/CONTRIBUTING.md.
-->

### Link to issue number:
Fixes https://github.com/nvaccess/nvda/issues/9288

### Summary of the issue:
PDF 2.0 added Associated Files (AF). It also describes a method for Formula tags to make use of AF that contain MathML. The LaTeX Project (the group that maintains LaTeX) has released an update to LaTeX that uses this technique. Hence, there will soon be a large body of PDF documents generated from LaTeX (pdflatex and lualatex) that contain MathML. 

In conjunction with Foxit and an informal agreement with someone at Adobe, we agreed on a method to expose the MathML in an AF without a change to the PDF accessibility interface: the Formula tag gets role=Math (in windows, ROLE_SYSTEM_EQUATION) and the contents of tag is the MathML.

Note: this does not change the legality of the previous method of fully tagging the PDF math with children elements pointing to subexpressions in the PDF. However, that method has proved difficult to implement for PDF generators. This method seems to be much simpler and hence will be used.

The latest release of Foxit contains the support of AF with MathML. So far, Adobe has not made a change but with Foxit and NVDA supporting this, there will be more of an impetuses to do so. According to the Foxit implementer, it only took 1-2 days to implement. 

### Description of user facing changes
The math in documents will be spoken and brailled just as it is done for HTML documents. It will also be navigable. This should work with any of the MathML add-ons.

### Description of development approach
Support required only about 3 lines to be added to the AdobeAcobat.py file. I changed a few more lines to add debug warnings when various COM interfaces were not found.

There was a commit in January 2024 that wiped out the MathML support in PDF in favor of alt text. This was in the .cpp file that is part of this PR. This PR mostly reverts that change. Alt text is still supported via the creation of a MathML `<mtext>` element. Potentially, this is a better solution because sometimes the alt text is LaTeX and LaTeX contains lots of punctuation characters that are not spoken by NVDA by default. Pushing this to the Math handler gives them the ability to override this behavior and speak all the characters. Currently MathCAT just passes the `mtext` content directly to NVDA, but I will look into making it smarter about that.

Because Adobe Reader currently does not handle AFs, the alt text will get read if a formula has both an AF and alt text.

### Testing strategy:
Here are two PDF files for testing:
1. [Several inline and display equations](https://github.com/user-attachments/files/17334945/mathml-AF-ex2.pdf)
2. [Some equations with alt text](https://github.com/user-attachments/files/17334946/formula-alt-text.pdf)

### Known issues with pull request:
None

### Code Review Checklist:
I need some guidance on what to do, if anything, about "change log entry" and unit or system tests.
<!--
This checklist is a reminder of things commonly forgotten in a new PR.
Authors, please do a self-review of this pull-request.
Check items to confirm you have thought about the relevance of the item.
Where items are missing (eg unit / system tests), please explain in the PR.
To check an item `- [ ]` becomes `- [x]`, note spacing.
You can also check the checkboxes after the PR is created.
A detailed explanation of this checklist is available here:
https://github.com/nvaccess/nvda/blob/master/projectDocs/dev/githubPullRequestTemplateExplanationAndExamples.md#code-review-checklist
-->

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

<!-- Please keep the following -->
@coderabbitai summary

NSoiffer committed:
Mostly revert commit #fbfcd14b5f14609fba7a919c67431a1912b0da51 ("Adobe Reader: support alternative text on formulas in PDFs (#16067)")

This change doesn't allow MathML in formula tags. That was ok before, but now pdftex puts out MathML in an associated file and FoxIt reader makes it the content of a formula tag.

Adobe is generally on board with this idea but hasn't implemented handling of associated files yet. Currently Adobe Reader does not know about Associated Files. So the contents are the alternative text. This is handled by making it an `mtext` element.

Note: if the alt text is LaTeX, then it has lots of "punctuation" such as `{` and `\`. This typically would _not_ be read by NVDA. MathCAT currently passes this along unchanged to NVDA, but I have marked this as a bug and will have MathCAT produce speech that won't be discarded, so the experience will be an improved one over just speaking the alt text.

pre-commit-ci[bot] committed:
Pre-commit auto-fix

ABuffEr commented:
This sounds so beautiful!
@NSoiffer, thanks a lot to follow these evolutions. I ended my studies in 2019, and never had MathML materials; but it's very comforting to know that, in the future, I could find accessible PDF almost as easily as a sighted person.

codeofdusk commented:
@NSoiffer Just curious, is this new support part of the base `amspath` package or does it require an inclusion of an additional package at build time?

codeofdusk commented:
As for "change log entry", add a new item in `user_docs/en/changes.md` (similar to the others). Include the undelying issue number and your username as an at mention.

It looks like NV Access might need to manually start CI for this PR. Be sure that existing unit and system tests pass.

NSoiffer commented:
> @NSoiffer Just curious, is this new support part of the base amspath package or does it require an inclusion of an additional package at build time?

This is separate, although it does include support for the amsmath package. It is a project that has been worked on for 4(?) years, funded by Adobe that is nearing completion of phase V (their last listed phase). [Here is a relatively recent paper](https://www.latex-project.org/publications/2024-FMi-DPC-UFi-JAW-doceng24.pdf) on their work. It (currently) requires adding some metadata at the start of the file:
```
\DocumentMetadata{
  lang        = en,
  pdfversion  = 2.0,
  pdfstandard = ua-2,
  pdfstandard = a-4f, %or a-4
  testphase   = 
   {phase-III,
    title,
    table,
    math,
    firstaid}  
}
```
The first part of the metadata will be required as it states some important things that go into PDF's metadata. I suspect the "testphase" part will not be required at some point in the future.

More info about the project is on [their web page](https://www.latex-project.org/) and also [in their repo](https://github.com/latex3/tagging-project?tab=readme-ov-file) (it is a little out of date -- a month ago or so they started automatically generating MathML from TeX).
```

NSoiffer committed:
Add MathML support in PDF change.

NSoiffer committed:
Merge branch 'SupportMathML' of github.com:NSoiffer/nvda into SupportMathML

pre-commit-ci[bot] committed:
Pre-commit auto-fix

NSoiffer commented:
@codeofdusk: thanks. I've added something to the userDoc file.

As for testing, the only code that I changed is PDF related. As far as I can find, there are no tests for PDF files. `runsystemtests.bat -i "xxx"` where "xxx" is "PDF", "Adobe", and "Acrobat" all come back with "Suite 'Robot' contains no tests matching tags". So no tests should have been broken.

Because the code does COM calls, I don't see a way to do Unit tests. I don't know how to do PDF system tests because one needs a PDF processor running (AdobeReader, Foxit Reader, ...). The need for a PDF processor probably makes system testing hard (impossible?). If someone can tell me how to do some tests, I'll add some. Otherwise, I think the checklist is comple.

NSoiffer commented:
Having not heard any suggestions about how I could do testing, I've checked the box off. That means there are no tasks left. I think this is ready for review.

seanbudd commented on source/NVDAObjects/IAccessible/adobeAcrobat.py:
```diff
@@ -137,15 +143,30 @@ def _getNodeMathMl(self, node):
+			raise LookupError
 		# There could be other stuff before the math element. Ug.
+		mathMl = self.pdDomNode.GetValue()
+		log.debug(
+			f'\n_get_mathMl: math recognized: {mathMl.startswith("<math")}, child count={self.pdDomNode.GetChildCount()}',
+		)
+		log.debug(f"\nname={self.pdDomNode.GetName()}\nvalue={self.pdDomNode.GetValue()}")
```
I don't think a new line should be added here, won't this just create an unusual new line split between the log time and thread prefix and the log message.
If you want to group the log messages and have new lines, please join the messages together

```suggestion
		log.debug(
			f'\n_get_mathMl: math recognized: {mathMl.startswith("<math")}, child count={self.pdDomNode.GetChildCount()}',
			f"\nname={self.pdDomNode.GetName()}\nvalue={self.pdDomNode.GetValue()}"
		)
```

seanbudd commented on source/NVDAObjects/IAccessible/adobeAcrobat.py:
```diff
@@ -137,15 +143,30 @@ def _getNodeMathMl(self, node):
 		yield "</%s>" % tag
 
 	def _get_mathMl(self):
```
```suggestion
	def _get_mathMl(self) -> str:
```

seanbudd commented on source/NVDAObjects/IAccessible/adobeAcrobat.py:
```diff
@@ -137,15 +143,30 @@ def _getNodeMathMl(self, node):
 			except COMError:
+				log.debugWarning(f"COMError trying to get childNum={childNum}")
 				continue
+			log.debug(f"  get_mathMl: tag={child.GetTagName()}")
```
```suggestion
			log.debug(f"\tget_mathMl: tag={child.GetTagName()}")
```

seanbudd commented on user_docs/en/changes.md:
```diff
@@ -5,7 +5,7 @@
 
 ### New Features
-
+* Support for math in PDFs has been added. This works for formulas that have MathML in an Associated File. Tagged formulas are supported in newer versions of TeX/LaTeX (9288, @NSoiffer)
```
Please preserve the new line, use new lines for sentences, and use a # before the issue number

```suggestion

* Support for math in PDFs has been added.
This works for formulas that have MathML in an Associated File.
Tagged formulas are supported in newer versions of TeX/LaTeX (#9288, @NSoiffer)
```

SaschaCowley commented on source/NVDAObjects/IAccessible/adobeAcrobat.py:
```diff
@@ -137,15 +143,30 @@ def _getNodeMathMl(self, node):
 		# There could be other stuff before the math element. Ug.
+		mathMl = self.pdDomNode.GetValue()
+		log.debug(
+			f'\n_get_mathMl: math recognized: {mathMl.startswith("<math")}, child count={self.pdDomNode.GetChildCount()}',
```
Why are we querying the same thing multiple times here and below?

seanbudd replied:
I would also consider wrapping debug queries in `if log.isEnabledFor(log.DEBUG):` for performance

seanbudd reviewed (approved):
Thanks @NSoiffer this will be a much appreciated fix

SaschaCowley commented on source/NVDAObjects/IAccessible/adobeAcrobat.py:
```diff
@@ -137,15 +143,30 @@ def _getNodeMathMl(self, node):
+		log.debug(f"\nname={self.pdDomNode.GetName()}\nvalue={self.pdDomNode.GetValue()}")
+		# this test and the replacement doesn't work if someone uses a namespace tag (which they shouldn't, but..)
+		if mathMl.startswith("<math"):
+			return mathMl.replace('xmlns:mml="http://www.w3.org/1998/Math/MathML"', "")
```
What happens if this is `False`?

NSoiffer replied:
If `mathMl.startswith("<math")` is false, it moves on to see if there is "fully tagged" MathML. 

Fully tagged PDF math is where each subexpression is tagged with a child MathML element (e.g., `mfrac`) and points to the subpart of the rendered expression. It turns out this is very hard to generate, so there are very few documents in the world that do that. It is however legal and mentioned in the PDF/UA-2 spec. If it doesn't find any children, there may be alt text as the content. That is what the last line does. If there isn't content, then it is an empty `<mtext>` element, which is legal.

With FoxIt Reader, it there is MathML in an associated file, the test will be true. Since Adobe Reader doesn't yet deal with associated files, the test will be false, it won't find any children, and alt text will be returned by the last line.

SaschaCowley replied:
Thanks for the thorough explanation @NSoiffer. Also, apologies for the rather silly question. I was out of it yesterday, and somehow missed the `return`.

seanbudd committed:
Apply suggestions from code review

pre-commit-ci[bot] committed:
Pre-commit auto-fix

SaschaCowley commented on user_docs/en/changes.md:
```diff
@@ -5,7 +5,7 @@
 
 ### New Features
-
+* Support for math in PDFs has been added. This works for formulas that have MathML in an Associated File. Tagged formulas are supported in newer versions of TeX/LaTeX (9288, @NSoiffer)
```
Can we add that this requires a supported PDF client?

NSoiffer replied:
Will do.

codeofdusk replied:
Also, write one sentence per line for the translation system.

NSoiffer committed:
Clean up comments and log.debug statements

NSoiffer committed:
Improve the change comment

Added info about PDF processors that support this.

NSoiffer committed:
Merge branch 'SupportMathML' of github.com:NSoiffer/nvda into SupportMathML

NSoiffer committed:
Made each sentence be on its own line as requested by @codeofdusk.

NSoiffer committed:
Fix multiline debug "fix" -- actually tested the code this time...

pre-commit-ci[bot] committed:
Pre-commit auto-fix

SaschaCowley commented on source/NVDAObjects/IAccessible/adobeAcrobat.py:
```diff
@@ -136,16 +142,37 @@ def _getNodeMathMl(self, node):
 
-	def _get_mathMl(self):
-		# There could be other stuff before the math element. Ug.
+	def _get_mathMl(self) -> str:
```
Can this get a docstring please?

SaschaCowley reviewed (changes requested):
Thanks @NSoiffer. Just a few small documentation requests.

SaschaCowley committed:
Update user_docs/en/changes.md

cary-rowen commented:
To be honest I wouldn't be opposed to documenting some technical details on how to use this feature in the user guide. Even if they look somewhat professional.
Overall, those who wish to use a screen reader to read mathematical content will probably always need to know some technical knowledge.

NSoiffer committed:
Add documentation string.

Also, make a comment a bit clearer.

NSoiffer committed:
Merge branch 'SupportMathML' of github.com:NSoiffer/nvda into SupportMathML

NSoiffer committed:
Update user_docs/en/changes.md

Co-authored-by: Sascha Cowley <16543535+SaschaCowley@users.noreply.github.com>

NSoiffer committed:
Merge branch 'master' into SupportMathML

seanbudd commented on user_docs/en/changes.md:
```diff
@@ -7,11 +7,16 @@
 * Support for math in PDFs has been added.
-This works for formulas that have MathML in an Associated File.
-Tagged formulas are supported in newer versions of TeX/LaTeX (#9288, @NSoiffer)
+This works for Formulas with associated MathML mark-up, such as some files generated by newer versions of TeX/LaTeX.
```
```suggestion
This works for formulas with associated MathML mark-up, such as some files generated by newer versions of TeX/LaTeX.
```

NSoiffer committed:
Update user_docs/en/changes.md

Co-authored-by: Sean Budd <seanbudd123@gmail.com>

XLTechie commented on user_docs/en/changes.md:
```diff
@@ -6,6 +6,12 @@
 ### New Features
 
+* Support for math in PDFs has been added.
+This works for formulas with associated MathML mark-up, such as some files generated by newer versions of TeX/LaTeX.
```
Isn't "MathML mark-up" somewhat redundant? "Math Mark-up Language mark-up"

```suggestion
This works for formulas with associated MathML, such as some files generated by newer versions of TeX/LaTeX.
```

NSoiffer replied:
Yes, it is redundant. It is a bit like saying "the attached method is Python code".  Both "code" and "markup" are not needed, but seem to be more common. But I'm too close to MathML, so I defer to others and am fine with your suggestion.

NSoiffer committed:
Update user_docs/en/changes.md

Co-authored-by: Luke Davis <8139760+XLTechie@users.noreply.github.com>

NSoiffer commented:
GitHub says that @SaschaCowley requested updates. I must be missing something -- as far as I can see there are no unresolved requests. Can someone clarify for me what is not resolved?

Thanks.

SaschaCowley commented:
> GitHub says that @SaschaCowley requested updates. I must be missing something -- as far as I can see there are no unresolved requests. Can someone clarify for me what is not resolved?
> 
> Thanks.

This is just because I have not yet submitted an approving review. My apologies for the delay, I've been off sick for the last four days.

SaschaCowley commented on user_docs/en/changes.md:
```diff
@@ -6,6 +6,12 @@
+* Support for math in PDFs has been added.
+This works for formulas with associated MathML, such as some files generated by newer versions of TeX/LaTeX.
+Currently this is only supported in Foxit Reader & Foxit Editor. (#9288, @NSoiffer)
+* The volume of other applications can be adjusted by `NVDA+alt+pageUp` and `NVDA+alt+pageDown`. In order to use this feature, application volume adjuster needs to be enabled in Audio pane of NVDA settings. (#16052, @mltony)
+* Added command to mute or unmute all other applications, assigned to `NVDA+alt+delete`.
+In order to use this feature, the application volume adjuster needs to be enabled in the Audio category of NVDA settings. (#16052, @mltony)
```
```suggestion
```

SaschaCowley committed:
Update user_docs/en/changes.md
